plutus-core/testlib/PlutusIR/Generators/QuickCheck/GenerateTerms.hs

```diff
@@ -0,0 +1,853 @@
+      [ (lamF, genLam a b)    | Just (a, b) <- [funTypeView mty] ] ++
+      [ (varAppF, genVarApp mty) ]                                 ++
+      [ (10, genApp mty) ]                                         ++
+      [ (1, genError mty) ]                                        ++
+      [ (10, genConst mty)    | canConst mty ]                     ++
+      [ (10, genDatLet mty)   | YesEscape <- [esc] ]               ++
+      [ (10, genIfTrace)      | isNothing mty ]                    ++
+      [ (customF, customG mty) ]
+  where
+
+    checkInvariants gen = do
+      (ty, tm) <- gen
+      debug    <- asks geDebug
+      tyctx    <- asks geTypes
+      tmctx    <- asks geTerms
+      if debug then
+        case typeCheckTermInContext tyctx tmctx tm ty of
+          Left err ->
+             (error . show $ "genTerm - checkInvariants: term " <> prettyPirReadable tm
+                           <> " does not type check at type " <> prettyPirReadable ty
+                           <> " in type context " <> prettyPirReadable tyctx
+                           <> " and term context " <> prettyPirReadable tmctx
+                           <> " with error message " <> fromString err)
+          _ -> return (ty, tm)
+      else
+        return (ty, tm)
+
+    funTypeView Nothing                             = Just (Nothing, Nothing)
+    funTypeView (Just (normalizeTy -> TyFun _ a b)) = Just (Just a, Just b)
+    funTypeView _                                   = Nothing
+
+    -- Generate builtin ifthenelse and trace calls
+    genIfTrace = do
+      fun <- liftGen $ elements [IfThenElse, Trace]
+      pure (typeOfBuiltinFunction def fun, Builtin () fun)
+
+    genError Nothing = do
+      ty <- genType $ Type ()
+      return (ty, Error () ty)
+    genError (Just ty) = return (ty, Error () ty)
+
+    canConst Nothing            = True
+    canConst (Just TyBuiltin{}) = True
+    canConst (Just _)           = False
+
+    genConst Nothing = do
+      b <- deliver . liftGen . genBuiltinTypeOf $ Type ()
+      (TyBuiltin () b, ) <$> genConstant b
+    genConst (Just ty@(TyBuiltin _ b)) = (ty,) <$> genConstant b
+    genConst _ = error "genConst: impossible"
+
+    genDatLet mty = do
+      rec <- liftGen arbitrary
+      genDatatypeLet rec $ \ dat -> do
+        (ty, tm) <- genTerm mty
+        return $ (ty, Let () (if rec then Rec else NonRec) (DatatypeBind () dat :| []) tm)
+
+    genLet mty = do
+      -- How many terms to bind
+      n   <- liftGen $ choose (1, 3)
+      -- Names of the bound terms
+      xs  <- genLikelyFreshNames $ replicate n "f"
+      -- Types of the bound terms
+      -- TODO: generate something that matches the target type
+      as  <- onAstSize (`div` 8) $ vectorOf n $ genType (Type ())
+      -- Strictness
+      ss  <- vectorOf n $ liftGen $ elements [Strict, NonStrict]
+      -- Recursive?
+      r   <- liftGen $ frequency [(5, pure True), (30, pure False)]
```

Comment:
`5/30` because that's easier to tweak than `1/6` when you've got a whole bunch of these things lying around. In this case probably just a leftover from lots of different tweaking in lots of different places.

It's the other way around, we avoid generating excessively many recursive types.

Reply:
> It's the other way around, we avoid generating excessively many recursive types.

I've no idea how I got it backwards :facepalm: 

Thanks!